File: 2.Data_Structure_Implementations/Tree_Based_Data_Structures/Binary_Tree/Python/binary_tree.py
from queue import Queue

class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None

    def insert(self, value):
        q = Queue()
        q.put(self)

        while not q.empty():
            current = q.get()

            if not current.left:
                current.left = BinaryTree(value)
                break
            elif not current.right:
                current.right = BinaryTree(value)
                break
           
            q.put(current.left)
            q.put(current.right)

        return self

    def contains(self, value):
        q = Queue()
        q.put(self)

        # Searches with level order traversal
        while not q.empty():
            current = q.get()

            if current.value == value:
                return True

            if current.left:
                 q.put(current.left)
            if current.right:
                q.put(current.right)
        
        return False

    def remove(self, value):
        if self is None:
            return None

        # single-node tree, root deletion
        if self.__is_leaf_node(self) and self.value == value:
            self.value = None
            return self

        node_to_delete = None
        parent = None
        last_node = None

        q = Queue()
        q.put(self)

        while not q.empty():
            # after loop is this this will be the deepest
            # node in tree
            last_node = q.get()

            if last_node.value == value:
                node_to_delete = last_node

            if last_node.left:
                parent = last_node
                q.put(last_node.left)
            if last_node.right:
                parent = last_node
                q.put(last_node.right)

        
        if node_to_delete:
            # replace values
            node_to_delete.value = last_node.value

            # remove deepest node from the tree
            if parent.right is last_node:
                parent.right = None
            else:
                parent.left = None

        return self

    # Helper Methods

    def __is_leaf_node(self, node):
        return not node.left and not node.right

    def print_tree(self, level=0, prefix="Root: "):
        print(" " * (level * 4) + prefix + str(self.value))
        if self.left:
            self.left.print_tree(level + 1, "L--- ")
        if self.right:
            self.right.print_tree(level + 1, "R--- ")

File: 2.Data_Structure_Implementations/Tree_Based_Data_Structures/Binary_Tree/JavaScript/binaryTree.js
/**
 * Binary tree implementation with a queue to achieve O(1) enqueue & dequeue,
 * instead of the O(n) runtime given by the array implementation
 */
class BinaryTree {
    constructor(value) {
        this.value = value
        this.right = null
        this.left = null
    }

    insert(value) {
        const q = new Queue()
        q.enqueue(this)

        while (!q.isEmpty()) {
            const current = q.dequeue()

            if (!current.left) {
                current.left = new BinaryTree(value)
                break
            } else if (!current.right) {
                current.right = new BinaryTree(value)
                break
            }
            q.enqueue(current.left)
            q.enqueue(current.right)
        }
        return this
    }

    contains(value) {
        const q = new Queue()
        q.enqueue(this)

        while (!q.isEmpty()) {
            const current = q.dequeue()

            if (current.value === value)
                return true

            if (current.left)
                q.enqueue(current.left)

            if (current.right)
                q.enqueue(current.right)
        }
        return false
    }

    remove(value) {
        // Check if it's a single-node tree
        if (this.#isLeafNode(this) && this.value == value) {
            this.value = null
            return this
        }

        let nodeToDelete = null
        let lastNodeParent = null
        let lastNode = null

        const q = new Queue()
        q.enqueue(this)

        while (!q.isEmpty()) {
            // after loop is this will be the deepest
            // node in the tree
            lastNode = q.dequeue()

            if (lastNode.value === value)
                nodeToDelete = lastNode


            if (lastNode.left) {
                lastNodeParent = lastNode
                q.enqueue(lastNode.left)
            }
            if (lastNode.right) {
                lastNodeParent = lastNode
                q.enqueue(lastNode.right)
            }

        }

        if (nodeToDelete) {
            // swap node values
            nodeToDelete.value = lastNode.value

            // remove deepest node from the tree
            if (lastNodeParent.right === lastNode)
                lastNodeParent.right = null
            else
                lastNodeParent.left = null
        }
        return this
    }

    // Private method to check if node is a leaf node in the tree
    #isLeafNode(node) {
        return !node.left && !node.right
    }

    printTree(level = 0, prefix = "Root: ") {
        console.log(" ".repeat(level * 4) + prefix + this.value)
        if (this.left)
            this.left.printTree(level + 1, "L--- ")

        if (this.right)
            this.right.printTree(level + 1, "R--- ")
    }
}

// Queue implementation since JavaScript does not hava a built in data structure
class Queue {
    #front
    #back
    #size

    constructor() {
        this.#front = null
        this.#back = null
        this.#size = 0
    }

    enqueue(data) {
        if (this.isEmpty()) {
            this.#front = this.#back = new Node(data)

            this.#size = 1
            return true
        }

        const newNode = new Node(data)
        this.#back.next = newNode
        this.#back = newNode
        this.#size++
        return true
    }

    dequeue() {
        if (this.isEmpty())
            return null

        const deletedNode = this.#front
        this.#front = this.#front.next

        // If queue becomes empty, reset back
        if (this.#front == null)
            this.#back = null

        this.#size--
        return deletedNode.data
    }

    peek() {
        return this.isEmpty() ? null : this.#front.data;
    }

    // ====================
    //   Utility Methods
    // ====================

    getSize() {
        return this.#size
    }

    isEmpty() {
        return this.getSize() == 0
    }
}

class Node {
    constructor(data) {
        this.data = data
        this.next = null
    }
}